ompi/mca/coll/ucc/coll_ucc_allgather.c

```diff
@@ -0,0 +1,106 @@
+        goto fallback;
+    }
+
+    opal_datatype_type_size(&sdtype->super, &dt_size);
+    ucc_coll_args_t coll = {
+        .mask      = 0,
+        .coll_type = UCC_COLL_TYPE_ALLGATHER,
+        .src.info = {
+            .buffer   = (void*)sbuf,
+            .count    = scount,
+            .datatype = ucc_sdt,
+            .mem_type = UCC_MEMORY_TYPE_UNKNOWN
+        },
+        .dst.info = {
+            .buffer   = (void*)rbuf,
+            .count    = rcount * ucc_module->ucc_team->size,
+            .datatype = ucc_rdt,
+            .mem_type = UCC_MEMORY_TYPE_UNKNOWN
+        }
+    };
+
+    if (MPI_IN_PLACE == sbuf) {
+        coll.mask  = UCC_COLL_ARGS_FIELD_FLAGS;
+        coll.flags = UCC_COLL_ARGS_FLAG_IN_PLACE;
+    }
+    COLL_UCC_REQ_INIT(coll_req, req, coll, ucc_module);
+    return UCC_OK;
+fallback:
+    return UCC_ERR_NOT_SUPPORTED;
+}
+
+int mca_coll_ucc_allgather(const void *sbuf, int scount, struct ompi_datatype_t *sdtype,
+                           void* rbuf, int rcount, struct ompi_datatype_t *rdtype,
+                           struct ompi_communicator_t *comm,
+                           mca_coll_base_module_t *module)
+{
+    mca_coll_ucc_module_t *ucc_module = (mca_coll_ucc_module_t*)module;
+    ucc_coll_req_h         req;
+
+    UCC_VERBOSE(3, "running ucc allgather");
+    COLL_UCC_CHECK(mca_coll_ucc_allgather_init(sbuf, scount, sdtype,
+                                              rbuf, rcount, rdtype,
+                                              ucc_module, &req, NULL));
+    COLL_UCC_CHECK(ucc_collective_post(req));
+    COLL_UCC_CHECK(coll_ucc_req_wait(req));
+    return OMPI_SUCCESS;
+fallback:
+    UCC_VERBOSE(3, "running fallback allgather");
+    return ucc_module->previous_allgather(sbuf, scount, sdtype, rbuf, rcount, rdtype,
+                                          comm, ucc_module->previous_allgather_module);
+}
+
+int mca_coll_ucc_iallgather(const void *sbuf, int scount, struct ompi_datatype_t *sdtype,
+                            void* rbuf, int rcount, struct ompi_datatype_t *rdtype,
+                            struct ompi_communicator_t *comm,
+                            ompi_request_t** request,
+                            mca_coll_base_module_t *module)
+{
+    mca_coll_ucc_module_t *ucc_module = (mca_coll_ucc_module_t*)module;
+    ucc_coll_req_h         req;
+    mca_coll_ucc_req_t    *coll_req;
+
+    UCC_VERBOSE(3, "running ucc iallgather");
+    COLL_UCC_GET_REQ(coll_req);
+    COLL_UCC_CHECK(mca_coll_ucc_allgather_init(sbuf, scount, sdtype,
+                                              rbuf, rcount, rdtype,
+                                              ucc_module, &req, coll_req));
+    COLL_UCC_CHECK(ucc_collective_post(req));
+    *request = &coll_req->super;
+    return OMPI_SUCCESS;
+fallback:
+    UCC_VERBOSE(3, "running fallback iallgather");
+    return ucc_module->previous_iallgather(sbuf, scount, sdtype, rbuf, rcount, rdtype,
```

Comment:
we have a leak of coll_req here (it was in the original coll/ucc implementation). If we go to fallback then coll_req might leak. Please set coll_req to NULL in line 92. Then before line 104 check:
if (coll_req) {
mca_coll_ucc_req_free(coll_req);
}
"previous_iallgather" will allocate the request itself.

Also please check & apply the same fix to ALL other non-blocking coll/ucc collectives